### PR TITLE
Container runtime env support

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Fixes the container runtime env plugin + adds tests

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by jjyao — COMMENTED** [review]
Do we have tests?

**Comment by zcin** [comment]
> Do we have tests?

@jjyao I'll be adding a test to the Kuberay repo for this (will link here when I have it up). To add a test to the Ray CI, the docker container that we run tests in needs to be started with privileged permissions.

File: python/ray/_private/runtime_env/container.py
Comment (rkooo567) [inline python/ray/_private/runtime_env/container.py:37]: should we make this configurable? or root is necessary?

File: python/ray/_private/runtime_env/container.py
Comment (rkooo567) [inline python/ray/_private/runtime_env/container.py:35]: Q: is shm shared across all containers?

File: python/ray/_private/runtime_env/context.py
Comment (rkooo567) [inline python/ray/_private/runtime_env/context.py:76]: I don't understand why it is needed here. can you comment why here?

Comment (zcin) [reply python/ray/_private/runtime_env/context.py:76]: By default, raylet uses the path to `default_worker.py` on host. However, the path to `default_worker.py` inside the container started by podman can be different. This is how the original API works, but the `worker_path` value was being ignored in code; this part fixes it.

File: python/ray/_private/runtime_env/container.py
Comment (zcin) [reply python/ray/_private/runtime_env/container.py:35]: I believe it shares memory through the option `--ipc=host` (although I'm not super familiar with the details)

File: python/ray/_private/runtime_env/container.py
Comment (zcin) [reply python/ray/_private/runtime_env/container.py:37]: There are some files in the mounted folder `/tmp/ray` that the `default_worker.py` process needs to access inside the container, so it complains if I don't run the container as root because of permissions issues. Not sure how this worked originally without root user, this was originally written 2 years ago and there were no tests, so a lot of things in core probably changed during that time.

Happy to dig into it more if you have time for a pair session. For now, I added this because it doesn't work without it.

**Comment by jjyao** [comment]
> @jjyao I'll be adding a test to the Kuberay repo for this (will link here when I have it up). To add a test to the Ray CI, the docker container that we run tests in needs to be started with privileged permissions. I think when this was originally contributed, no tests were enabled in the CI.

I think we should try to add tests now (here instead of kuberay repo) if we want to move this plugin out of experimental/alpha.

**Comment by zcin** [comment]
> I think we should try to add tests now (here instead of kuberay repo) if we want to move this plugin out of experimental/alpha.

Agree that if we want to move this out of experimental/alpha, that is necessary. In Serve we are planning to document this as an experimental feature though, since it definitely isn't stable yet from what I've tested. Do you think we need to add the test if it remains experimental?

**Comment by jjyao** [comment]
Yea, I think we should still add tests.

File: ci/docker/serve.build.Dockerfile
Comment (aslonnie) [inline ci/docker/serve.build.Dockerfile:20]: 1. please add the deb source first, and then update only ones, and then install all packages they you need in one line.
2. move these before the `COPY . .`

and we are using buildkit, so consider using a multiline. see the other exising `RUN <<EOF` for example. multiline will generate fewer image layers and is faster to load.

File: ci/ray_ci/container.py
Comment (aslonnie) [inline ci/ray_ci/container.py:96]: please make this a flag so that it is enabled by default for everything.

least privilege principle.

File: python/ray/_private/runtime_env/container.py
Comment (rkooo567) [reply python/ray/_private/runtime_env/container.py:35]: hmm we may need tests.

File: ci/ray_ci/container.py
Comment (can-anyscale) [reply ci/ray_ci/container.py:96]: the flag can be passed down through this script https://github.com/ray-project/ray/blob/master/ci/ray_ci/tester.py#L40; poke me if you need any help ;)

File: ci/docker/serve.build.Dockerfile
Comment (can-anyscale) [reply ci/docker/serve.build.Dockerfile:20]: Let's fix this

**Review by can-anyscale — APPROVED** [review]
CI changes look good! Just need to fix @aslonnie's comments before merging. Thanks!

File: python/ray/_private/runtime_env/container.py
Comment (alexeykudinkin) [reply python/ray/_private/runtime_env/container.py:35]: With `--ipc=host` containers will have access to host's shared memory (ie created outside of any container on the host itself)

File: python/ray/_private/runtime_env/container.py
Comment (alexeykudinkin) [reply python/ray/_private/runtime_env/container.py:37]: We'd try to avoid elevating permissions to root

Which would mean we need to carefully manage permissions of things in the `/tmp/ray`

File: python/ray/_private/runtime_env/context.py
Comment (alexeykudinkin) [reply python/ray/_private/runtime_env/context.py:76]: Can we instead of overriding an argument so deep in the call-stack do it where invoke the worker instead?

File: python/ray/serve/tests/test_deployment_in_container.py
Comment (alexeykudinkin) [inline python/ray/serve/tests/test_deployment_in_container.py:24]: Are these public Docker images (we need to make sure there are no credentials leaked in either of them)?

File: python/ray/serve/tests/test_deployment_in_container.py
Comment (alexeykudinkin) [inline python/ray/serve/tests/test_deployment_in_container.py:66]: Can clean up

File: python/ray/serve/tests/test_deployment_in_container.py
Comment (alexeykudinkin) [inline python/ray/serve/tests/test_deployment_in_container.py:77]: Can we validate these tests fail if we remove `--ipc=host`?

File: python/ray/serve/tests/test_deployment_in_container.py
Comment (alexeykudinkin) [inline python/ray/serve/tests/test_deployment_in_container.py:126]: Can you help me understand what we're testing with this test?

Comment (zcin) [reply python/ray/serve/tests/test_deployment_in_container.py:126]: This checks that the log files exists for a worker that's started in its own container.

File: python/ray/_private/runtime_env/context.py
Comment (zcin) [reply python/ray/_private/runtime_env/context.py:76]: I think this is the location where we invoke the worker: https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/context.py#L85-L95

File: python/ray/serve/tests/test_deployment_in_container.py
Comment (zcin) [reply python/ray/serve/tests/test_deployment_in_container.py:24]: These simply use `rayproject/ray:nightly-py38-cpu` as the base image, downgrade python version to 3.8.16, then write "hello world" to a file for testing.

**Review by rkooo567 — COMMENTED** [review]
One main concern is how it works when it is used with other runtime envs. (e.g., env_vars).

Can you give me how each other runtime env API will work when they are provided with a container plugin?

File: python/ray/_private/runtime_env/context.py
Comment (zcin) [reply python/ray/_private/runtime_env/context.py:71]: This is the worker path that the user sets in runtime_env["container"]["worker_path"]. If the user doesn't set it, it is None.

File: python/ray/serve/_private/application_state.py
Comment (zcin) [reply python/ray/serve/_private/application_state.py:407]: The application state manager lives in the controller, and this should only be called anytime a new config is deployed in Serve.

**Comment by rkooo567** [comment]
initial decision: we disallow every other runtime env other than container when container runtime env is used. We will follow up upon user requests

File: docker/container-runtime-env-tests/Dockerfile
Comment (zcin) [reply docker/container-runtime-env-tests/Dockerfile:3]: This is no longer used, will be removed

File: ci/build/build-ray-docker.sh
Comment (aslonnie) [inline ci/build/build-ray-docker.sh:14]: as discussed, please revert these change. rather, please add an additional layer on top of ray image to install podman.

File: ci/ray_ci/builder.py
Comment (aslonnie) [inline ci/ray_ci/builder.py:71]: please revert.

File: ci/ray_ci/docker_container.py
Comment (aslonnie) [inline ci/ray_ci/docker_container.py:18]: please revert.

File: python/ray/tests/conftest_docker.py
Comment (aslonnie) [inline python/ray/tests/conftest_docker.py:184]: this is going to be always the same as the group id of `/var/run/docker.sock` that is mounted in the test env. you do not have to get it inside the container.

you can also run `stat -c %g /var/run/docker.sock` instead of using regex parsing on the ls output, which can be easy to make mistakes.

File: python/ray/tests/conftest_docker.py
Comment (aslonnie) [inline python/ray/tests/conftest_docker.py:191]: what is this `daemon` group?

File: python/ray/tests/conftest_docker.py
Comment (aslonnie) [inline python/ray/tests/conftest_docker.py:212]: nit: use a for loop over a list of list, so that you do not need to repeat the `container_id`?

File: python/ray/tests/conftest_docker.py
Comment (aslonnie) [inline python/ray/tests/conftest_docker.py:219]: why `check_output` when you do not need the output? maybe just `check_call`?

File: python/ray/tests/test_runtime_env_container.py
Comment (aslonnie) [inline python/ray/tests/test_runtime_env_container.py:45]: not particular a fan of these inlined python files.. maybe just put them into a directory and build them into your container image with meaningful names?

and then you can run `python <filename.py>` rather than `-c`.

File: python/ray/tests/test_runtime_env_container.py
Comment (aslonnie) [inline python/ray/tests/test_runtime_env_container.py:500]: nit: remove these ones that are in comments?

File: python/ray/tests/BUILD
Comment (aslonnie) [inline python/ray/tests/BUILD:471]: nit: could you format this file properly with `buildifier`?

Comment (zcin) [reply python/ray/tests/BUILD:471]: I've never used `buildifier` before, how can I do that?

**Comment by zcin** [comment]
Hi @pcmoritz @thomasdesr could one of you review the changes to the telemetry files?

File: python/ray/tests/conftest_docker.py
Comment (zcin) [reply python/ray/tests/conftest_docker.py]: This is unnecessary and was used for testing, I've removed it

File: python/ray/tests/conftest_docker.py
Comment (zcin) [reply python/ray/tests/conftest_docker.py]: I've switched to `stat -c %g /var/run/docker.sock`, thanks for the pointer!

File: python/ray/tests/BUILD
Comment (zcin) [reply python/ray/tests/BUILD:471]: I've formatted this build rule to match the indentation around it!

File: docker/runtime_env_container/Dockerfile
Comment (aslonnie) [inline docker/runtime_env_container/Dockerfile:11]: could you only `apt-get update` once?

pressumably:

1. add deb source
2. apt update
3. apt install all packages you need

**Review by aslonnie — APPROVED** [review]
the CI part looks good to me.

I leave the ray runtime env implementation and test parts for code owners to review.

thanks for the refactoring on the test env setup. great job!

File: docker/runtime_env_container/Dockerfile
Comment (zcin) [reply docker/runtime_env_container/Dockerfile:11]: So I tried removing the apt-get update at the beginning, and it can't seem to find the packages `software-properties-common` and `uidmap`, not sure why 😕

File: python/ray/_private/runtime_env/container.py
Comment (rkooo567) [inline python/ray/_private/runtime_env/container.py:44]: Consider adding a link if there is any!

File: python/ray/_private/runtime_env/container.py
Comment (rkooo567) [inline python/ray/_private/runtime_env/container.py:50]: Add comments these env is needed for the default worker

File: python/ray/_private/runtime_env/container.py
Comment (rkooo567) [inline python/ray/_private/runtime_env/container.py:52]: So, we are syncing with all env vars starting with RAY_ right? 

Can you add an unit test for this?

File: python/ray/tests/runtime_env_container/test_worker_exit_intended_system_exit_and_user_error.py
Comment (rkooo567) [inline python/ray/tests/runtime_env_container/test_worker_exit_intended_system_exit_and_user_error.py:127]: Q: why do these files need to be in a separate file like this?

Comment (rkooo567) [reply python/ray/tests/runtime_env_container/test_worker_exit_intended_system_exit_and_user_error.py:127]: Can you comment the reasoning to `test_runtime_env_container.py`?

File: python/ray/tests/test_runtime_env_container.py
Comment (rkooo567) [inline python/ray/tests/test_runtime_env_container.py:87]: Assert on error messages as well!

File: python/ray/tests/test_runtime_env_container.py
Comment (rkooo567) [inline python/ray/tests/test_runtime_env_container.py:85]:
```suggestion
class TestContainerRuntimeEnvWithOtherRuntimeEnv:
```
?

File: python/ray/tests/test_runtime_env_container.py
Comment (rkooo567) [reply python/ray/tests/test_runtime_env_container.py:87]: same for all tests below

**Review by rkooo567 — COMMENTED** [review]
Looks pretty good! Only nit comments. I think we can merge it after addressing the last comments

File: docker/runtime_env_container/Dockerfile
Comment (aslonnie) [reply docker/runtime_env_container/Dockerfile:11]: first, always use `apt-get` not `apt` in docker files. `apt` is not for humanless scenarios.

then, the thing I am asking you to do is:

````
    sudo sh -c "echo 'deb http://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04/ /' > /etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list" && \
    sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 4D64390375060AA4 && \
    sudo apt-get update && \
    sudo apt-get install -y software-properties-common uidmap podman
````

the reason is that:

apt-get update takes some time and there is no reason to do it twice. you want to add a new deb source so that you can install `podman`, and that should be done before `apt-get update`

I actually do not quite understand why `software-properties-common` and `uidmap` are required here.

Comment (edoakes) [reply docker/runtime_env_container/Dockerfile:11]: @aslonnie for my own understanding, why is `apt-get` preferred over `apt` for humanless scenarios (and vice versa)?

File: python/ray/tests/runtime_env_container/test_worker_exit_intended_system_exit_and_user_error.py
Comment (zcin) [reply python/ray/tests/runtime_env_container/test_worker_exit_intended_system_exit_and_user_error.py:127]: Hi Sang, there is a comment at the top of the `test_runtime_env_container.py` explaining the reasoning. It is to avoid having long inline python files that we run with `python -c`.

**Comment by rkooo567** [comment]
Lmk when all comments are addressed! I am going to approve the PR

**Comment by zcin** [comment]
> Lmk when all comments are addressed! I am going to approve the PR

@rkooo567 I've addressed your comments, please take another look!

**Review by rkooo567 — APPROVED** [review]

File: docker/runtime_env_container/Dockerfile
Comment (aslonnie) [reply docker/runtime_env_container/Dockerfile]: > why is apt-get preferred over apt for humanless scenarios (and vice versa)?

https://askubuntu.com/questions/990823/apt-gives-unstable-cli-interface-warning

APT shows the warning itself when being used on a non interactive terminal.

from `apt`'s manpage:

> The [apt](https://manpages.ubuntu.com/manpages/xenial/man8/apt.8.html)(8) commandline is designed as an end-user tool and it may change behavior between
       versions. While it tries not to break backward compatibility this is not guaranteed either
       if a change seems beneficial for interactive use.

**Review by pcmoritz — APPROVED** [review]
Approval for the usage.proto changes :)

**Comment by zcin** [comment]
@rkooo567 Tests are passing, is this ready to merge?

**Comment by SongGuyang** [comment]
Nice fix!

**Pull request merged**
